go/vt/vttablet/tabletmanager/semisyncmonitor/monitor.go

```diff
@@ -0,0 +1,390 @@
+// also checks that the write count value should not exceed
+// the maximum value configured. It returns whether it was able
+// to increment the value or not.
+func (m *Monitor) incrementWriteCount() bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.inProgressWriteCount == maxWritesPermitted {
+		return false
+	}
+	m.inProgressWriteCount = m.inProgressWriteCount + 1
+	m.writesBlockedGauge.Set(int64(m.inProgressWriteCount))
+	return true
+}
+
+// AllWritesBlocked returns if maxWritesPermitted number of writes
+// are already outstanding.
+func (m *Monitor) AllWritesBlocked() bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.isOpen && m.inProgressWriteCount == maxWritesPermitted
+}
+
+// decrementWriteCount decrements the write count.
+func (m *Monitor) decrementWriteCount() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.inProgressWriteCount = m.inProgressWriteCount - 1
+	m.writesBlockedGauge.Set(int64(m.inProgressWriteCount))
+}
+
+// write writes a heartbeat to unblock semi-sync being stuck.
+func (m *Monitor) write() {
+	shouldWrite := m.incrementWriteCount()
+	if !shouldWrite {
+		return
+	}
+	defer m.decrementWriteCount()
+	// Get a connection from the pool
+	conn, err := m.appPool.Get(context.Background())
+	if err != nil {
+		log.Errorf("SemiSync Monitor: failed to write to semisync_recovery table: %v", err)
+		return
+	}
+	defer conn.Recycle()
+	_, err = conn.Conn.ExecuteFetch(m.bindSideCarDBName(semiSyncRecoverWrite), 0, false)
+	if err != nil {
+		log.Errorf("SemiSync Monitor: failed to write to semisync_recovery table: %v", err)
+	}
+}
+
+// setIsBlocked sets the isBlocked field.
+func (m *Monitor) setIsBlocked(val bool) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.isBlocked = val
+	if !val {
+		// If we are unblocked, then we need to signal all the waiters.
+		for _, ch := range m.waiters {
+			close(ch)
+		}
+		// We also empty the list of current waiters.
+		m.waiters = nil
+	}
+}
+
+// clearAllData clears all the data in the table so that it never
+// consumes too much space on the MySQL instance.
+func (m *Monitor) clearAllData() {
+	// Get a connection from the pool
+	conn, err := m.appPool.Get(context.Background())
+	if err != nil {
+		log.Errorf("SemiSync Monitor: failed to clear semisync_recovery table: %v", err)
+		return
+	}
+	defer conn.Recycle()
+	_, err = conn.Conn.ExecuteFetch(m.bindSideCarDBName(semiSyncRecoverClear), 0, false)
+	if err != nil {
+		log.Errorf("SemiSync Monitor: failed to clear semisync_recovery table: %v", err)
+	}
+}
+
+// addWaiter adds a waiter to the list of waiters
+// that will be unblocked when the primary is no longer blocked.
+func (m *Monitor) addWaiter() chan any {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	ch := make(chan any)
+	m.waiters = append(m.waiters, ch)
+	return ch
+}
+
+// bindSideCarDBName binds the sidecar db name to the query.
+func (m *Monitor) bindSideCarDBName(query string) string {
+	return sqlparser.BuildParsedQuery(query, sidecar.GetIdentifier()).Query
+}
```

Comment:
The sidecar is immutable in a tablet instance so we could potentially do this once.

Reply:
Yeah, fine to leave it as-is. I was just thinking and saying it. 🙂